web/src/compose_validate.ts

```diff
@@ -795,3 +796,92 @@ export function convert_mentions_to_silent_in_direct_messages(
     const silent_mention_text = people.get_mention_syntax(full_name, user_id, true);
     return silent_mention_text;
 }
+
+export const SPLIT_DELIMITER = "\n\n\n";
+
+function delimiter_index_outside_code_block(message_content: string): number {
+    let index = message_content.indexOf(SPLIT_DELIMITER);
+    while (index !== -1) {
+        if (
+            !compose_ui.position_inside_code_block(message_content, index + 1) &&
+            !compose_ui.position_inside_code_block(message_content, index + 2)
+        ) {
+            return index;
+        }
+        index = message_content.indexOf(SPLIT_DELIMITER, index + 1);
+    }
+    return -1;
+}
+/**
+ * Splits a message into 2 parts, if possible
+ * @param message_content the content to be split
+ * @param split_messages whether the option is enabled
+ * @returns a pair of strings: the first being the individual split message, the second being the
+ * remaining part of the message, which will be undefined when there are no further parts to send
+ */
+export function split_message(
+    message_content: string,
+    split_messages: boolean,
+): [string, string | undefined] {
+    if (!split_messages) {
+        return [message_content, undefined];
+    }
+    if (!message_content) {
+        return ["", undefined];
+    }
+    const index = delimiter_index_outside_code_block(message_content);
+    if (index === -1) {
+        return [message_content, undefined];
+    }
+    const send_content = message_content.slice(0, index);
+    const rest_content = message_content.slice(index + SPLIT_DELIMITER.length);
+    if (send_content.length === 0) {
+        // there are too many "\n\n\n" sequences, so some parts are empty strings
+        // so we find the next non-empty part
+        return split_message(rest_content, true);
+    }
+    return [send_content, rest_content];
+}
+
+/**
+ * @param content the content to be split
+ * @returns a list of non-empty message parts, split by the delimiter
+ */
+export function count_message_content_split_parts(
+    content = compose_state.message_content(),
+): number {
+    let count = 1;
+    let parts = split_message(content, true);
+    while (parts[1]) {
+        count += 1;
+        parts = split_message(parts[1], true);
+    }
+    return count;
+}
+
+/**
+ * checks whether the message actually needs to be split, as it is possible
+ * that `split_messages` is enabled but the message doesn't contain the delimiter.
+ */
+export function will_split_into_multiple_messages(): boolean {
+    if (!is_split_messages_enabled()) {
+        return false;
+    }
+    return count_message_content_split_parts() > 1;
+}
+
+const ls = localstorage();
+let split_messages_enabled = ls.get("split_messages") === "true";
```

Comment:
I don't understand why we need this variable at all, and it anyway should have been defined at the top, not right before a function. Reading and setting it directly in local storage should suffice?

Reply:
I used the variable in case localstorage is not supported. I'm then manipulating the variable and the value in the localstorage separately so that we can at least enable/disable this setting even if localstorage is unavailable.

Reply:
I think this warrants a discussion on CZO... You can also check how local storage is used in the codebase for other features for reference

Reply:
I've initiated a discussion on CZO. 
I've split the changes into several commits which should be easy to review. 
Apart from the syncing issue, everything else is as per the issue description and the discussions on CZO.